### PR TITLE
Add `strategy` argument to `StreamResponse.enable_compression()` method

### PR DESCRIPTION
**Pull request by shootkin**
## What do these changes do?

This PR adds `strategy` argument to StreamResponse.enable_compression() method.
For example, tests show that download speed becomes 2x faster for `.png` images when using
`zlib.Z_RLE` compresson strategy, like
```
response.enable_compression(strategy=zlib.Z_RLE)
```
## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: CHANGES/6257.feature
Comment (webknjaz) [inline CHANGES/6257.feature]: This shouldn't follow the Git commit message rules. Other principles apply for the changelog. Also, it's RST and not Markdown which is why linters and RTD fail.

https://github.com/aio-libs/aiohttp/tree/master/CHANGES#readme

Comment (webknjaz) [reply CHANGES/6257.feature]: Also, this sentence focuses on the implementation details which are not of interest to the release notes readers. Instead try showing what this enables people to do.

File: docs/web_reference.rst
Comment (webknjaz) [inline docs/web_reference.rst]:
```suggestion
      *strategy* accepts a :mod:`zlib` compression strategy.
```

File: docs/web_reference.rst
Comment (webknjaz) [inline docs/web_reference.rst]:
```suggestion
      Possible values are :data:`zlib.Z_DEFAULT_STRATEGY`,
```

Comment (webknjaz) [reply docs/web_reference.rst]: (whether it's really "data" should be possible to check via intersphinx-untangled)

**Comment by webknjaz** [comment]
Oh, the deps are broken, that's unrelated.

File: docs/web_reference.rst
Comment (shootkin) [reply docs/web_reference.rst]: Sorry, I can't find whether it is data or not here: https://webknjaz.github.io/intersphinx-untangled/docs.python.org/
zlib.Z_DEFAULT_STRATEGY is of type int, so should I use :data: here or not?

File: CHANGES/6257.feature
Comment (shootkin) [reply CHANGES/6257.feature]: Changed the description to appropriate format

File: docs/web_reference.rst
Comment (shootkin) [reply docs/web_reference.rst]: Fixed

**Comment by webknjaz** [comment]
@shootkin your commits don't match your GitHub profile. You need to fix your local `git config` setup with the proper email for this to work.

File: aiohttp/web_response.py
Comment (Dreamsorcerer) [inline aiohttp/web_response.py:331]: As before, pretty sure this is going to cause a type error, which we'll see once the CI is running again.

Maybe @asvetlov can answer the best way to proceed. Will need 1 of:

- Change `self._payload_writer` to `StreamWriter`.
- Add this parameter to `AbstractStreamWriter`.
- Use an `isinstance(StreamWriter, ...)` check before using the parameter.

File: aiohttp/web_response.py
Comment (Dreamsorcerer) [inline aiohttp/web_response.py]:
```suggestion
        self._compression_strategy: Optional[int] = None
```

Comments for types are from legacy code, let's not add any more.

**Review by Dreamsorcerer — CHANGES_REQUESTED** [review]

**Comment by shootkin** [comment]
> @shootkin your commits don't match your GitHub profile. You need to fix your local `git config` setup with the proper email for this to work.

Just changed email

File: aiohttp/web_response.py
Comment (shootkin) [reply aiohttp/web_response.py:331]: @webknjaz What do you think about it?

**Comment by Dreamsorcerer** [comment]
If you merge master now, then the checks will run properly.

**Comment by shootkin** [comment]
> If you merge master now, then the checks will run properly.

Sorry, merge which way? I cannot merge it to master because it **is** master

**Comment by Dreamsorcerer** [comment]
> > If you merge master now, then the checks will run properly.
> 
> Sorry, merge which way? I cannot merge it to master because it **is** master

aio-libs:master into your branch.

**Review by shootkin — COMMENTED** [review]
Add strategy argument to StreamResponse.enable_compression() method

File: aiohttp/web_response.py
Comment (shootkin) [reply aiohttp/web_response.py:331]: Added `if isinstance(self._payload_writer, StreamWriter)` condition

File: aiohttp/web_response.py
Comment (Dreamsorcerer) [inline aiohttp/web_response.py]: Seems like `None` is not a valid strategy. Why not set it to the same default (`zlib.Z_DEFAULT_STRATEGY`)?

Otherwise, maybe we could reuse the `self._compression` property. Define it as `Optional[int]` and then change the check in `_prepare_headers()` to `if self._compression is not None:`. Might need a little more tweaking to get it right though.

File: docs/web_reference.rst
Comment (webknjaz) [reply docs/web_reference.rst]: Yes, use it but also add to `nitpick_ignore` in `docs/conf.py`.

File: aiohttp/web_response.py
Comment (shootkin) [reply aiohttp/web_response.py]: I've changed `self._compression_strategy` from `None` to `zlib.Z_DEFAULT_STRATEGY` but it didn't help.
Reuse `self._compression`? But how it is related to `self._compression_strategy`?
I changed `strategy` parameter in `StreamWriter.enable_compression()` from `strategy: int = zlib.Z_DEFAULT_STRATEGY` to `strategy: Optional[int] = zlib.Z_DEFAULT_STRATEGY` and it did the trick.

File: docs/web_reference.rst
Comment (shootkin) [reply docs/web_reference.rst]: Done

File: aiohttp/web_response.py
Comment (shootkin) [reply aiohttp/web_response.py]: Sorry, `make lint` failed again. Finally I changed `Optional[int]` to `int` everywhere because linter says that `zlib.compressobj()` does not accept `Optional[int]`.
Now `mypy` test is being passed successfully for my changes

**Review by shootkin — COMMENTED** [review]
Please review

File: aiohttp/web_response.py
Comment (Dreamsorcerer) [inline aiohttp/web_response.py]: Tests are passing now. I'd still like @asvetlov or someone to confirm if we should use the `isinstance()` check (or change the type to `StreamWriter` or add the parameter to `AbstractStreamWriter`) before merging.

Comment (Dreamsorcerer) [reply aiohttp/web_response.py]: From what I can see, the abstract class is not used in our code. It is only a StreamWriter within our internal code, so I don't even see a way for anyone to replace it. So, I suspect the abstract is not really needed. For now, let's just change the abstract class.

File: aiohttp/abc.py
Comment (Dreamsorcerer) [inline aiohttp/abc.py]:
```suggestion
        self, encoding: str = "deflate", strategy: int = zlib.Z_DEFAULT_STRATEGY
```

File: aiohttp/web_response.py
Comment (Dreamsorcerer) [reply aiohttp/web_response.py]:
```suggestion
            self._payload_writer.enable_compression(
                coding.value, self._compression_strategy
            )
```

File: docs/web_reference.rst
Comment (Dreamsorcerer) [inline docs/web_reference.rst]:
```suggestion
      See :func:`zlib.compressobj` for possible values.
```

File: docs/conf.py
Comment (Dreamsorcerer) [inline docs/conf.py]:
```suggestion
```

File: aiohttp/web_response.py
Comment (Dreamsorcerer) [inline aiohttp/web_response.py]:
```suggestion
```

**Pull request merged**
